### ISSUE TITLE
blake2 not found when python is built with `--without-builtin-hashlib-hashes`

### ISSUE THREAD

**Issue by sshedi** [python/cpython#95835]
# Bug report
I have built python-3.9 with `--without-builtin-hashlib-hashes` option enabled and when I do `import hashlib` with built binary, I'm seeing errors.
```
$ ./python -c "import hashlib"
ERROR:root:code for hash blake2b was not found.
Traceback (most recent call last):
  File "/root/Python-3.9.1/Lib/hashlib.py", line 254, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/root/Python-3.9.1/Lib/hashlib.py", line 129, in __get_openssl_constructor
    return __get_builtin_constructor(name)
  File "/root/Python-3.9.1/Lib/hashlib.py", line 123, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type blake2b
ERROR:root:code for hash blake2s was not found.
Traceback (most recent call last):
  File "/root/Python-3.9.1/Lib/hashlib.py", line 254, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/root/Python-3.9.1/Lib/hashlib.py", line 129, in __get_openssl_constructor
    return __get_builtin_constructor(name)
  File "/root/Python-3.9.1/Lib/hashlib.py", line 123, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type blake2s
```

# Your environment
- CPython versions tested on: 3.9
- Operating system and architecture: Photon OS (dev branch), amd64
- openssl: 3.0.3

openssl is built with blake2 support but still seeing this error. Am I doing something wrong or this is how it is designed?
```
$ openssl dgst -list
Supported digests:
-blake2b512                -blake2s256                -md4
-md5                       -md5-sha1                  -mdc2
-ripemd                    -ripemd160                 -rmd160
-sha1                      -sha224                    -sha256
-sha3-224                  -sha3-256                  -sha3-384
-sha3-512                  -sha384                    -sha512
-sha512-224                -sha512-256                -shake128
-shake256                  -sm3                       -ssl3-md5
-ssl3-sha1                 -whirlpool
```

**Comment by sshedi** [comment]
cc: @tiran

**Comment by tiran** [comment]
OpenSSL's blake2 API is not sufficient for Python. It lacks several features,

https://github.com/python/cpython/blob/d348afa15d5a997e7a8e51c0f789f41cb15cc651/Lib/hashlib.py#L73-L80

**Comment by sshedi** [comment]
Thanks for the quick response @tiran.
How can I solve this issue? Do I need to use blake2 from hashlib and disable other hashlib builtin modules?

**Comment by sshedi** [comment]
And any inputs on this error? I have enabled both kernel & openssl fips in my system. This basically makes `hashlib` unusable on fips enabled machines, right?

``` 
$ ./python -c "import hashlib"
ERROR:root:code for hash md5 was not found.
Traceback (most recent call last):
  File "/root/Python-3.9.1/Lib/hashlib.py", line 137, in __get_openssl_constructor
    f(usedforsecurity=False)
ValueError: [digital envelope routines] unsupported

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/Python-3.9.1/Lib/hashlib.py", line 254, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/root/Python-3.9.1/Lib/hashlib.py", line 141, in __get_openssl_constructor
    return __get_builtin_constructor(name)
  File "/root/Python-3.9.1/Lib/hashlib.py", line 123, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type md5
```

**Comment by tiran** [comment]
Blake2 is not permitted in FIPS mode anyway, so you have to leave it disabled and cannot use it.

Python does not fully support OpenSSL 3.0 yet. Further Python 3.9 is lacking several FIPS-related fixes that are required for OpenSSL 3.0 in FIPS mode. Since Python 3.9 is in security fix mode, we cannot backport any of the OpenSSL changes.

Hint: You may want to check how CentOS 9 has implemented FIPS support for Python 3.9.

**Comment by sshedi** [comment]
Okay, thanks for the inputs. I will check them out.
